[충남대 FE 김은수] 2단계 상품 상세 페이지 API 구현하기

shushuburger opened the pull request:
안녕하세요! 충남대학교 FE 12팀 김은수입니다.

### 과제의 요구사항 및 수행내용

- [x] 본인만의 기준으로 일관된 코드를 작성해주세요.
  - [x] hook들이 한 파일에 너무 많은 역할을 하고 있지는 않은지, 상수 처리할 수 있는 부분이 있는지 등을 점검했습니다.
  - [x] 하나의 컴포넌트에 너무 많은 요소들이 들어가지는 않았는지 분리할 수 있는 컴포넌트가 있지는 않은지를 기준으로 삼아 ProductDetailPage에 있던 요소들을 여러 컴포넌트로 나누어주었습늬다.
  - [x] prop drilling이 될수도 있을 것 같아 ProductDetailInfo에서 일괄적으로 진행하던 API 요청을 하위 단계 컴포넌트들에서 각각 본인에게 필요한 요청만 하도록 로직을 분리하여 작성해주었습니다.
- [x] 기능 단위로 나누어 커밋을 해주세요.
- [x] 상품 상세 페이지
  - [x] [참고 URL](https://kakaotech-mission4-api-advanced.pages.dev/) 을 참고하여 상품 상세 페이지를 구현해주세요.
    - [x] ProductDetailPage.tsx를 만들어 상품 상세 정보 페이지를 위한 틀을 생성해두었습니다.
    - [x] 로그인이 되어있지 않으면 로그인 후에 상품 상세 정보 페이지에 접근할 수 있도록 App.tsx에서 PrivateRoute 안에 ProductDetailPage로 이동하는 Route를 추가해주었습니다.
    - [x] routes.ts에 url 관련 변수를 상수로 추가해주었습니다.
    - [x] 화면에 표시해야 하는 모든 요소들을 api를 통해 성공적으로 받아오는 것을 확인했습니다.
    - [x] 기본적인 UI 작업을 완료했습니다.
    - [x] "주문하기" 버튼을 누르면 OrderPage로 이동하도록 해주었습니다.
    - [x] 상품 카드를 누르면 ProductDetailPage로 연결되도록 해주었습니다.
  - [x] API 통신의 최적화, 선언적 구조, react query 등을 활용하여 기능을 구현하세요.
  - [x] 아래 참고 API를 모두 사용해야 해요.
    - [x] 상품 정보 API: **`/api/products/:productId`**
      - [x] useProductInfo.ts를 만들어 해당 api에 대한 요청을 처리하는 로직을 작성했습니다.
      - [x] api.ts에 api 경로와 관련된 변수를 상수 처리해주었습니다.
    - [x] 상품 세부 정보 API: **`/api/products/:productId/detail`**
      - [x] useProductDetail.ts를 만들어 해당 api에 대한 요청을 처리하는 로직을 작성했습니다.
      - [x] api.ts에 api 경로와 관련된 변수를 상수 처리해주었습니다.
    - [x] 상품 주요 리뷰 API: **`/api/products/:productId/highlight-review`**
      - [x] useHighlightReview.ts를 만들어 해당 api에 대한 요청을 처리하는 로직을 작성했습니다.
      - [x] api.ts에 api 경로와 관련된 변수를 상수 처리해주었습니다.
    - [x] 상품 관심 등록 수 API: **`/api/products/:productId/wish`**
      - [x] useProductWish.ts를 만들어 해당 api에 대한 요청을 처리하는 로직을 작성했습니다.
      - [x] api.ts에 api 경로와 관련된 변수를 상수 처리해주었습니다.
  - [x] 상품 관심 등록 버튼 클릭 시 낙관적 업데이트를 통해 상품 관심 등록 수를 변경해보세요. (실제 API 반영은 없기 때문에, 새로고침 하면 사라지는 것이 정상이에요.)
    - [x] handleWishClick를 통해서 "찜 버튼"을 누르면 하트가 채워지고 위시 수가 +1이 되는 로직을 다루도록 해주었습니다.
    - [x] useProductWish에 staleTime을 추가해주어서 일정 시간 동안은 서버에 post하지 않았더라도 isWished가 반영된 상태를 유지하도록 해주었습니다.
- [x] ErrorBoundary와 Suspense를 사용하여 코드 구조를 리팩터링 해요. (단, ErrorBoundary는 관련 라이브러리를 사용하지 않고 구현해요)
  - [x] useSuspenseQuery를 사용해서 기존 isError, isPending으로 다루던 코드를 대체해주었습니다.
  - [x] 이번에 새로 작업한 "상품 상세 정보 화면"에 관련한 hook, component들에도 suspense + ErrorBoundary 작업을 진행해주었습니다.
- [x] npm run build 성공하기
- [x] npm run lint 성공하기

shushuburger committed:
docs(readme): 2단계 구현 필수 내용 표시

shushuburger committed:
feat(route): 상품 상세 페이지 라우트 추가

- /products/:productId 경로를 ProductDetailPage.tsx로 연결
- useParams를 통해 productId 추출
- 로그인 필요 페이지로 설정하여 PrivateRoute로 보호
- README에 구현 내용 표시

shushuburger committed:
fix(rotue): url 경로 오타 수정

shushuburger committed:
feat(api): 상품 기본 정보 API 경로 및 useProductInfo 훅 연동

- getProductInfoPath, getProductInfoUrl 유틸 함수 추가
- 상품 상세 API 호출 시 baseUrl 적용
- useProductInfo 훅 내부 axios 요청을 유틸로 추출
- README에 구현 내용 표시

shushuburger committed:
fix(alias): import 절대 경로로 수정

shushuburger committed:
feat(api): useProductDetail 훅 구현 및 상품 설명/공지사항 연동

- /api/products/:productId/detail 호출을 위한 useProductDetail 훅 구현
- 상품 설명과 공지사항 데이터 패칭
- ProductDetailPage에서 해당 훅 적용하여 상세 정보 표시
- dangerouslySetInnerHTML로 HTML 설명 출력 처리
- README에 구현 내용 표시

shushuburger committed:
feat(api): useProductWish 훅 구현

- /api/products/:productId/wish API 호출을 위한 useProductWish 훅 정의
- 상품 찜 수와 찜 여부 조회
- 상품 상세 페이지에서 찜 정보 확인 테스트
- API 유틸 getProductWishUrl로 경로 구성 분리
- README에 구현 내용 표시

shushuburger committed:
chore(debug): useProductDetail, useProductWish 훅 내부 console.log 제거

- 개발 중 사용했던 응답 로그 출력 코드 제거

shushuburger committed:
feat(api): useHighlightReview 훅 구현

- /api/products/:productId/highlight-review API 연동
- 리뷰 총 개수와 리뷰 배열 조회
- react-query 기반으로 useHighlightReview 훅 정의
- API 유틸 함수 getHighlightReviewUrl로 분리
- README에 구현 내용 표시

shushuburger committed:
feat(detail): 상품 상세 페이지 UI 구조 및 탭 콘텐츠 구현

- ProductDetailInfo 컴포넌트 내 탭 구조(상품설명/선물후기/상세정보) 구현
- useParams로 productId를 받아 내부에서 useProductInfo, useProductDetail, useHighlightReview API 호출
- 상품명, 가격, 브랜드, 이미지 등 상단 정보 렌더링
- 상품설명 탭: description HTML 콘텐츠 표시
- 선물후기 탭: 작성자/내용 스타일링 포함 리스트 렌더링
- 상세정보 탭: 공지사항 항목명 + 본문 구조로 시각화

shushuburger committed:
feat(detail): 주문하기 버튼 클릭 시 해당 상품 주문 페이지로 이동 처리

- react-router-dom의 useNavigate 활용하여 주문 페이지로 이동
- 경로 이동 시 ROUTES.ORDER(productId) 상수 사용하여 경로 이동

shushuburger committed:
feat(route): 상품 클릭시 ProductDetailPage로 이동하도록 경로 수정

shushuburger committed:
feat(detail): 찜 버튼 낙관적 UI 업데이트 및 캐시 동기화 구현

- 상품 상세 하단 ProductActionBar에서 찜 버튼 클릭 시 wish 상태 및 count 즉시 반영
- useState로 wishCount, isWished 상태 관리
- useEffect를 통해 초기 wish 상태 동기화
- React Query의 queryClient.setQueryData 사용하여 productWish 캐시 수동 갱신
- staleTime 설정을 통해 페이지 재진입 시에도 하트 상태 유지되도록 처리
- 새로고침 시 서버 상태로 복구되는 동작 유지
- README에 구현 내용 표시

shushuburger committed:
style(detail): 상품 상세 화면 전체 스타일 정리 및 여백 개선

- 탭 버튼을 균등 3등분 처리해 전체 폭 활용
- 전체적인 margin, padding, gap 조정으로 실제 서비스와 유사한 밀도 구현
- README에 구현 내용 표시

shushuburger committed:
refactor(hooks): 상품 관련 API 훅 개선 및 에러 메시지 상수화

- useProductDetail 훅에서 하드코딩된 에러 메시지를 ERROR_MESSAGES.LOAD_PRODUCT_FAIL로 교체
- useProductInfo 훅에서도 동일한 방식으로 에러 메시지 상수화 적용
- useProductWish 훅에 ERROR_MESSAGES.FAILED_TO_LOAD_PRODUCT_WISH 사용
- useHighlightReview 훅에 ERROR_MESSAGES.FAILED_TO_LOAD_HIGHLIGHT_REVIEW 적용 및 변수명 개선
- README에 구현 내용 표시

shushuburger committed:
refactor(product): ProductDetailInfo 컴포넌트를 재사용성 기준으로 분리

- 이미지, 상품명, 가격, 브랜드 정보를 ProductHeader 컴포넌트로 분리
- 탭 선택 UI를 ProductTabSelector 컴포넌트로 분리하여 책임 분리
- PRODUCT_DETAIL_TABS 상수와 ProductDetailTab 타입을 constants로 추출
- ProductDetailInfo는 데이터 fetch 및 탭 전환 상태에 집중하도록 간결화
- README에 구현 내용 표시

shushuburger committed:
refactor(product): ProductDetailInfo 하위 컴포넌트에서 직접 API 호출하도록 리팩터링

- ProductHeader, ProductDescription, ProductReviews, ProductAnnouncements 컴포넌트 내부에서 productId 및 API 훅 직접 사용
- 상위 컴포넌트인 ProductDetailInfo에서는 API 상태와 데이터 전달 책임 제거
- prop drilling 제거로 컴포넌트 재사용성 및 독립성 향상시키려 시도
- README에 구현 내용 표시

shushuburger committed:
refactor(category): Suspense 기반 API 처리 구조로 리팩터링

- CategoryContent 내부에서 pending, error 분기 제거
- useCategoryThemes 훅을 useSuspenseQuery로 변경하여 타입 단언 제거 및 타입 안전성 확보
- CategoryGroup에서 ErrorBoundary + Suspense 조합 적용
- ErrorBoundary 내부 사용하지 않는 인자 제거하여 ESLint 오류 해결
- README에 구현 내용 표시

shushuburger committed:
refactor(product-detail): Suspense + ErrorBoundary 기반 구조로 전체 리팩터링

- useProductInfo, useProductDetail, useProductWish, useHighlightReview 훅을 useSuspenseQuery로 변경
- ProductDetailPage에 ErrorBoundary + Suspense 도입
- ProductDetailInfo 내부 탭 구조를 renderTab 함수로 정리
  - 각 탭 컴포넌트(ProductDescription, ProductReviews, ProductAnnouncements)에 Suspense + ErrorBoundary 적용
- 탭 컴포넌트 내부에서 isLoading, isError, !data 분기 제거
- ProductActionBar에서 로딩/에러 분기 제거 및 wish 상태 로컬 업데이트 유지
- README에 구현 내용 표시

shushuburger committed:
fix(layout): 스크롤 점프 현상 방지를 위해 html에 overflow-y: scroll 적용

- 콘텐츠 유무에 따라 스크롤바가 생기며 레이아웃이 밀리는 현상(scroll jump) 발생함을 인지
- html 태그에 overflow-y: scroll 설정하여 항상 스크롤 공간을 확보

shushuburger commented:
### Scroll Jump

첨부해주신 영상 덕분에 문제 상황을 저도 빨리 찾을 수 있었습니다! 감사합니다. 아마 영상이 없었더라면 더 많은 시간을 소비했을 것 같습니다 ㅎㅎ

"스크롤이 생겼다 사라졌다 하는 문제"이기에 해당 문제를 확인하려면 "스크롤이 없어도 되는 크기의 화면"을 우선적으로 띄워야겠다고 생각했습니다.
-> 화면 배율 조정으로 멘토님께서 남겨주신 영상과 동일한 현상을 확인

이에 대한 해결 방법을 고민하다가

> 같은 영역이지만 때에 따라 다른 요소들이 보일 때 높낮이가 마구 변해서 힘들었다
> -> 그렇다면 최소 높이나 너비를 지정해주어서 화면의 안정감을 확보하자!

와 같은 논리로 작업을 했던 기억이 났고 이번에도 이런 방향으로 접근해도 괜찮겠다고 결정을 내려서
기본적으로 "스크롤을 위한 공간"을 확보한 상태에서 모든 서비스가 화면에 띄워지도록 BaseStyle.ts에 
```
  html {
    overflow-y: scroll;
  }
```
코드를 추가해주었습니다.

이후 아래 영상처럼 화면이 바뀌어도 스크롤은 항상 고정되어 있는 결과를 만들었습니다

https://github.com/user-attachments/assets/ac28e3ac-aeac-41b3-9077-5f5b9e4daead

youngjewoo commented on src/components/ProductDetailSection/ProductDescription.tsx:
```diff
@@ -0,0 +1,28 @@
+  const { data: detail } = useProductDetail(productId!);
+
+  return (
+    <Description dangerouslySetInnerHTML={{ __html: detail.description }} />
```
[ 질문 ]

여기 `dangerouslySetInnerHTML` 이 속성을 꼭 사용할 필요가 있었나요..?? 그냥 div 에 텍스트 채워져도 될 것 같아서요!
프로퍼티 이름에서도 드러나듯이 사용하지 않을 수 있다면 dangerouslySetInnerHTML는 XSS 공격을 위한 경로로 사용되기가 쉬워서 사용하지 않는 것이 좋습니다!

shushuburger replied:
맨 처음에 <div>{detail.description}</div> 구조로 작성을 했을 때 HTML 문자열 그대로 화면에 출력되는 것을 확인했었습니다.
이로 인해 html 문자열을 화면에 element로 띄울 수 있는 방법!을 구글링해보았고
연관검색어로 인해 가장 먼저 본 것이 dangerouslySetInnerHTML이다보니 위험성을 무시한 채로 바로 사용했습니다.

피드백 이후에 더 안전한 방법을 탐구하다가 html-react-parser라는 
HTML 문자열을 React element로 안전하게 파싱해주는 외부 라이브러리가 있다는 것을 
새롭게 알게 되어서 이 라이브러리를 적용하면 되겠다!라고 생각했습니다.
최종적으로 html-react-parser로 대체해주었습니다!

앞으로는 새로운 개념을 도입할 때 위험성까지 더 고려해서 최대한 안전하고 유지보수가 쉬운 것을 처음부터 도입하려는 노력을 더 해보겠습니다!!

참고 링크: https://velog.io/@exceed96/React-HTML%EC%9A%94%EC%86%8C%EB%A5%BC-%ED%8F%AC%ED%95%A8%ED%95%9C-%EB%8D%B0%EC%9D%B4%ED%84%B0-%EB%A0%8C%EB%8D%94%EB%A7%81-%EB%B0%A9%EB%B2%95

youngjewoo replied:
아 API 응답이 HTML로 내려오는군요..! 이럴 때는 말씀하신 라이브러리 쓰셔도 되고요.
브라우저 내장 객체인 DOMParser를 통해 파싱하고 textContent만 꺼내서 사용하셔도 됩니다.

```ts
 const content = new DOMParser().parseFromString(`${HTML 컨텐츠}`, 'text/html')
```

꼭 바꾸셔야한다.. 까지는 아니고 참고요!

youngjewoo commented on src/components/ProductDetailSection/ProductDetailInfo.tsx:
```diff
@@ -0,0 +1,91 @@
+    switch (selectedTab) {
+      case '상품설명':
+        return (
+          <ErrorBoundary
```
[ change request ] 1. 코드 중복 제거

이 부분은 중복되는 부분을 분리하면 코드량이 훨씬 줄어들 수 있을 거 같은데요..! 다른 부분들은 다 잘 하셨으니 도전해보시길 바라요!

shushuburger replied:
ErrorBoundary+Suspense의 구조를 지니고 있는 부분을 common/WithSuspenseBoundary.tsx로 분리하여 주었고
해당 구조를 ProductDetailInfo 뿐만 아니라 ErrorBoundary + Suspense의 구조를 가진 곳에
일괄적으로 WithSuspenseBoundary를 적용하여 중복 코드를 줄이려 해보았습니다

youngjewoo commented on src/constants/validation.ts:
```diff
@@ -26,4 +26,7 @@ export const ERROR_MESSAGES = {
   NO_THEMES_AVAILABLE: '테마가 없습니다.',
   FAILED_TO_LOAD_PRODUCTS: '상품을 불러오지 못했어요.',
   NO_PRODUCTS_AVAILABLE: '상품이 없습니다.',
+  FAILED_TO_LOAD_PRODUCT_WISH: '상품 찜 정보를 불러오지 못했어요.',
```
이거는 미래를 위해서 넣어두신 거죠??

shushuburger replied:
기존에 ProductActionBar에서 사용하다가 ErrorBoundary로 묶으면서 "상품 찜"보다는 "상품"에 대한 정보를 불러오지 못했다고표시하는게 더 적합하다고 생각했습니다!
더이상 사용되지 않는 요소인데 제거하는 것을 잊었습니다.
제거하여 commit 하겠습니다!

youngjewoo reviewed (changes requested):
이전 단계까지 진행하시면서 코드 리뷰들을 잘 반영해주시고, 코드 정리도 잘 하면서 오셔서 이번 과제는 조금 수월하게 하셨을 거 같네요!

기능적 동작에 문제가 없어서 병합하려고 했으나..
약간의 코드 중복과, dangerouslysetinnerhtml 이 사용된 부분에 코멘트 드렸습니다.

해당 부분만 수정 혹은 답변 부탁드려요~~!
고생하셨습니다.

youngjewoo commented:
> ### Scroll Jump
> 첨부해주신 영상 덕분에 문제 상황을 저도 빨리 찾을 수 있었습니다! 감사합니다. 아마 영상이 없었더라면 더 많은 시간을 소비했을 것 같습니다 ㅎㅎ
> 
> "스크롤이 생겼다 사라졌다 하는 문제"이기에 해당 문제를 확인하려면 "스크롤이 없어도 되는 크기의 화면"을 우선적으로 띄워야겠다고 생각했습니다. -> 화면 배율 조정으로 멘토님께서 남겨주신 영상과 동일한 현상을 확인
> 
> 이에 대한 해결 방법을 고민하다가
> 
> > 같은 영역이지만 때에 따라 다른 요소들이 보일 때 높낮이가 마구 변해서 힘들었다
> > -> 그렇다면 최소 높이나 너비를 지정해주어서 화면의 안정감을 확보하자!
> 
> 와 같은 논리로 작업을 했던 기억이 났고 이번에도 이런 방향으로 접근해도 괜찮겠다고 결정을 내려서 기본적으로 "스크롤을 위한 공간"을 확보한 상태에서 모든 서비스가 화면에 띄워지도록 BaseStyle.ts에
> 
> ```
>   html {
>     overflow-y: scroll;
>   }
> ```
> 
> 코드를 추가해주었습니다.
> 
> 이후 아래 영상처럼 화면이 바뀌어도 스크롤은 항상 고정되어 있는 결과를 만들었습니다
> 
>  scroll.mp4

도움이 되셨다니 다행이네요.

과제에 명시된 내용은 아니었지만.. 사용자들에게는 체감이 되는 부분일 겁니다. 그래서 학습 과제로 전달드렸던 건데 능동적으로 문제 해결에 관심을 보여주셔서 저도 감사하네요! 실무에서는 보통 width는 화면 너비에 어느정도 가변 가능성을 항상 열어두기때문에 고정 너비를 보통은 잘 적용하지 않습니다. 반면 height의 경우는 모바일, 데스크톱에서 같은 height를 쓰는 경우도 있고, 부여하지 않으면 오히려 CLS등의 성능 문제나 UX가 좋지 않기때문에 고정 높이를 주로 부여하는 편이고요.

앞으로의 작업에 도움되시라고 남겨봅니다. 
이제 스크롤 점프가 발생하지 않네요. 고생하셨습니다~!

shushuburger committed:
refactor(common): TabPanel 컴포넌트 추출 및 공통 사용으로 Suspense + ErrorBoundary 중복 제거

- ErrorBoundary + Suspense 조합을 추상화한 TabPanel 컴포넌트를 common 폴더에 생성
- fallbackMessage props로 에러 메시지를 전달하여 유연하게 처리 가능하도록 구성
- ProductDetailInfo에서 탭별 컴포넌트를 TabPanel로 감싸도록 변경
- ProductDetailPage의 ProductDetailInfo + ProductActionBar 렌더링 영역에 TabPanel 적용
- CategoryGroup의 CategoryContent 영역에 TabPanel 적용
- 각 파일 내 중복된 fallback 텍스트 컴포넌트 및 loading import 제거
- 공통 처리로 유지보수성과 일관성 향상

shushuburger committed:
refactor(validation): 불필요한 에러 메시지 제거 및 문구 개선

- 사용되지 않는 ERROR_MESSAGES 항목 정리
- 상품 상제 정보 불러오지 못했다는 에러 메시지 추가

shushuburger committed:
refactor(product): ProductDescription에 html-react-parser 적용하여 HTML 파싱 처리

- dangerouslySetInnerHTML 대신 html-react-parser의 parse() 사용
- API에서 내려온 HTML 문자열을 React element로 안전하게 렌더링
- README에 구현 내용 표시

shushuburger committed:
refactor(common): TabPanel → WithSuspenseBoundary로 이름 변경

- 범용적인 ErrorBoundary + Suspense 조합 컴포넌트로 역할 명확히 표현
- TabPanel이라는 탭 전용 의미를 제거하고, 페이지 전역에서 사용할 수 있도록 개선
- WithSuspenseBoundary를 사용하는 곳 모두 import 경로 및 이름 변경

shushuburger committed:
refactor(common): TabPanel -> WithSuspenseBoundary로 이름 변경

shushuburger commented:
### 수정사항
- [x] 피드백 반영
  - [x] validation.ts에 사용하지 않는 변수 제거 및 조금 더 적절한 에러 메시지 문구 추가했습니다.
  - [x] ErrorBoundary + Suspense 구조를 WithSuspenseBoundary.tsx로 추출해 낸다음 기존에 ErrorBoundary + Suspense 구조를 활용하는 부분들을 WithSuspenseBoundary로 대체해주었습니다.
  - [x] html 구조로 제공되는 api 요청 결과를 화면에 성공적으로 띄우기 위해 기존의 dangerouslySetInnerHTML 대신 html-react-parser를 npm install로 설치하여 html 문자열을 element로 렌더링해 화면에 띄우도록 수정해주었습니다.
- [x] npm run build 성공하기
- [x] npm run lint 성공하기

youngjewoo commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -1,32 +1,20 @@
-import { Suspense } from 'react';
+// ProductDetailPage.tsx
```
요 주석은 없어도 되는 거죵?? ㅋ

shushuburger replied:
어머!! 지우고 commit을 했다고 생각했는데 반영이 안 되었었나 봅니다!!!! 다음 commit 때 불필요한 주석은 삭제하고 commit하겠습니다!

youngjewoo reviewed (approved):
HOC 개념도 알고 계시고 지난 시간동안 열심히 학습하신 티가 곳곳에서 나네요.

이번 PR은 병합하도록 하겠습니다! 고생하셨어요~~!
